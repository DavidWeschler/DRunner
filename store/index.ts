import { LocationStore, aiModelStore, MapThemeType, MarkerData, aiModel, hadasStore } from "@/types/type";
import { create } from "zustand";
import mapThemes from "../components/MapThemes/mapThemes";

export const useLocationStore = create<LocationStore>((set) => ({
  userLatitude: null,
  userLongitude: null,
  userAddress: null,
  destinationLatitude: null,
  destinationLongitude: null,
  destinationAddress: null,
  mapTheme: "standard",
  length: null,
  startPoint: null,
  startAddress: null,
  endPoint: null,
  endAddress: null,
  difficulty: null,
  inp: null,
  routeWayPoints: [], // might be depricated
  routeDirections: [], // might be depricated
  mode: "walking",
  routeDetalis: null,

  setRouteDetails: (route: any) => {
    set(() => ({
      routeDetalis: route,
    }));
  },

  setMode: (mode: string) => {
    set(() => ({
      mode,
    }));
  },

  setUserLocation: ({ latitude, longitude, address }: { latitude: number; longitude: number; address: string }) => {
    set(() => ({
      userLatitude: latitude,
      userLongitude: longitude,
      userAddress: address,
    }));
  },

  setDestinationLocation: ({ latitude, longitude, address }: { latitude: number; longitude: number; address: string }) => {
    set(() => ({
      destinationLatitude: latitude,
      destinationLongitude: longitude,
      destinationAddress: address,
    }));
  },

  setMapTheme: (theme: MapThemeType) => {
    set(() => ({
      mapTheme: theme, // Store only the theme name
    }));
  },

  setLengthInput: (length: number) => {
    set(() => ({
      length,
    }));
  },

  setStartPointInput: (point: { latitude: number; longitude: number } | null) => {
    set(() => ({
      startPoint: point ? { latitude: point.latitude, longitude: point.longitude } : null,
    }));
  },

  setStartAddress: (address: string) => {
    set(() => ({
      startAddress: address,
    }));
  },

  setEndPointInput: (point: { latitude: number; longitude: number } | null) => {
    set(() => ({
      endPoint: point ? { latitude: point.latitude, longitude: point.longitude } : null,
    }));
  },

  setEndAddress: (address: string) => {
    set(() => ({
      endAddress: address,
    }));
  },

  setDifficultyInput: (difficulty: string) => {
    set(() => ({
      difficulty,
    }));
  },

  setHadasInp: (inp: string) => {
    set(() => ({
      inp,
    }));
  },

  // setRouteWayPoints: (waypoints: MarkerData[]) => {
  //   set(() => ({
  //     routeWayPoints: waypoints,
  //   }));
  // },

  setRouteDirections: (directions: string[]) => {
    set(() => ({
      routeDirections: directions,
    }));
  },
}));

export const useaiModelStore = create<aiModelStore>((set) => ({
  model: { name: "Google Gemma 3", host: "google/gemma-3-4b-it:free" },
  setAiModel: (m: aiModel) => {
    set(() => ({
      model: m,
    }));
  },
}));

export const useHadasStore = create<hadasStore>((set) => ({
  chatReset: false,
  setChatReset: (state: boolean) => {
    set(() => ({
      chatReset: state,
    }));
  },
}));
