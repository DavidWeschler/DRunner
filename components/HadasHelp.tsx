import React, { useEffect, useState } from "react";
import { Modal, View, Text, TouchableOpacity, ScrollView, StyleSheet } from "react-native";
import { AboutModalProps } from "@/types/type";
import { useLocationStore, useaiModelStore } from "@/store";

const HadasHelp = ({ visible, onClose }: AboutModalProps) => {
  const { length, startAddress, endAddress, difficulty } = useLocationStore();
  const { model } = useaiModelStore();
  const [currentInputs, setCurrentInputs] = useState({
    "Running route length": length,
    "Start location": startAddress,
    "End location": endAddress,
    "Difficulty level": difficulty,
  });

  useEffect(() => {
    setCurrentInputs({
      "Running route length": length,
      "Start location": startAddress,
      "End location": endAddress,
      "Difficulty level": difficulty,
    });
  }, [length, startAddress, endAddress, difficulty]);

  return (
    <Modal animationType="slide" transparent visible={visible} onRequestClose={onClose}>
      <View style={styles.modalContainer}>
        <View style={styles.alertContainer}>
          <ScrollView contentContainerStyle={styles.scrollContainer}>
            <Text style={styles.title} className="text-md font-JakartaBold">
              How to use Hadas?
            </Text>
            <Text style={styles.bodyText}>Hadas is your AI assistant to help you plan your next running route or give you useful advice on how to create the best route for your needs!</Text>

            <Text style={styles.title} className="text-md font-JakartaBold">
              How to write to Hadas?
            </Text>
            <Text style={styles.bodyText}>You can ask Hadas in natural language, for example:</Text>
            <Text style={styles.exampleText}>
              - "I would like to run 3km starting from where I am and also end here. Make it easy."{"\n"}- "I'm not sure how long my running route should be. Any advice?"{"\n"}- "I haven't run in a while, should I go for a long run or a short one?"{"\n"}
            </Text>
            <Text style={styles.bodyText}>
              Hadas will ask you for more details if needed. The more details you provide, the better the recommendations will be!{"\n"}
              When the Generate button appears, you can see matching routes! üèÉ‚Äç‚ôÇÔ∏è‚Äç‚û°Ô∏è
            </Text>

            <Text style={[styles.title]} className="text-md font-JakartaBold">
              Current model: <Text className="text-md font-JakartaBold text-blue-700">{model.name}</Text>
            </Text>

            <Text style={styles.title} className="text-md font-JakartaBold">
              Current features:
            </Text>
            <View style={styles.currentInputsContainer}>
              {Object.entries(currentInputs).map(([key, value]) => (
                <View key={key} style={{ flexDirection: "row", marginBottom: 1 }}>
                  <Text style={styles.bodyText} className="text-md font-JakartaMedium text-gray-500">
                    {key}:
                  </Text>
                  <Text style={styles.bodyText} className="text-md font-JakartaBold text-blue-700">
                    {value ?? "Not set"}
                  </Text>
                </View>
              ))}
            </View>
          </ScrollView>

          <TouchableOpacity style={styles.closeButton} onPress={onClose} accessibilityRole="button">
            <Text style={styles.closeText}>Close</Text>
          </TouchableOpacity>
        </View>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "rgba(0, 0, 0, 0.5)",
  },
  alertContainer: {
    width: "85%",
    backgroundColor: "white",
    borderRadius: 12,
    padding: 20,
    alignItems: "center",
    position: "relative",
    overflow: "hidden",
  },
  scrollContainer: {
    paddingBottom: 20,
  },
  title: {
    fontSize: 18,
    marginBottom: 10,
    textAlign: "center",
  },
  bodyText: {
    fontSize: 14,
    marginBottom: 10,
    textAlign: "left",
  },
  exampleText: {
    fontSize: 14,
    fontStyle: "italic",
    color: "#555",
    marginBottom: 10,
  },
  currentInputsContainer: {
    padding: 15,
    borderRadius: 8,
    marginBottom: 2,
    width: "100%",
  },
  closeButton: {
    marginTop: 10,
    backgroundColor: "#0286FF",
    paddingVertical: 12,
    paddingHorizontal: 25,
    borderRadius: 8,
  },
  closeText: {
    color: "white",
    fontWeight: "bold",
    fontSize: 16,
  },
});

export default HadasHelp;
