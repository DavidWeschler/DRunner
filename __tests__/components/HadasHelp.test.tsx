import React from "react";
import { render } from "@testing-library/react-native";
import HadasHelp from "@/components/HadasHelp";
import { useLocationStore, useaiModelStore, useHadasStore } from "@/store";

// Mock store hooks
jest.mock("@/store", () => ({
  useLocationStore: jest.fn(),
  useaiModelStore: jest.fn(),
  useHadasStore: jest.fn(),
}));

// Mock CustomButton component
jest.mock("@/components/CustomButton", () => ({
  __esModule: true,
  default: ({ title, onPress }: { title: string; onPress: () => void }) => <button onClick={onPress}>{title}</button>,
}));

describe("HadasHelp Component", () => {
  const mockClose = jest.fn();
  const mockSetChatReset = jest.fn();

  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
    (useLocationStore as unknown as jest.Mock).mockReturnValue({
      length: 5,
      startAddress: "Test Start",
      endAddress: "Test End",
      difficulty: "Medium",
    });
    (useaiModelStore as unknown as jest.Mock).mockReturnValue({
      model: { name: "AI Model" },
    });
    (useHadasStore as unknown as jest.Mock).mockReturnValue({
      setChatReset: mockSetChatReset,
    });
  });

  it("renders correctly", () => {
    const { getByText, getByTestId } = render(<HadasHelp visible={true} onClose={mockClose} />);

    // Verify modal contents
    expect(getByText("How to use Hadas?")).toBeTruthy();
    expect(getByText("Current model: AI Model")).toBeTruthy();
    expect(getByText("5")).toBeTruthy();
    expect(getByText("Test Start")).toBeTruthy();
    expect(getByText("Test End")).toBeTruthy();
    expect(getByText("Medium")).toBeTruthy();
  });

  it("displays 'Not set' for empty inputs", () => {
    (useLocationStore as unknown as jest.Mock).mockReturnValue({
      length: 0,
      startAddress: "",
      endAddress: "",
      difficulty: "",
    });

    const { getByText } = render(<HadasHelp visible={true} onClose={mockClose} />);

    // Verify that 'Not set' is displayed for empty inputs
    expect(getByText("Running route length:")).toBeTruthy();
    expect(getByText("Start location:")).toBeTruthy();
    expect(getByText("End location:")).toBeTruthy();
    expect(getByText("Difficulty level:")).toBeTruthy();
  });

  it("matches snapshot", () => {
    const tree = render(<HadasHelp visible={true} onClose={mockClose} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
