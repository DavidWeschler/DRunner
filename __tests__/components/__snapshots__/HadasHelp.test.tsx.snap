// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HadasHelp Component matches snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.5)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 12,
          "height": "85%",
          "overflow": "hidden",
          "padding": 20,
          "position": "relative",
          "width": "85%",
        }
      }
    >
      <RCTScrollView
        contentContainerStyle={
          {
            "paddingBottom": 20,
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "fontFamily": "Jakarta-Bold",
                },
                {
                  "fontSize": 18,
                  "marginBottom": 10,
                  "textAlign": "center",
                },
              ]
            }
          >
            How to use Hadas?
          </Text>
          <Text
            style={
              {
                "fontSize": 14,
                "marginBottom": 10,
                "textAlign": "left",
              }
            }
          >
            Hadas is your AI assistant to help you plan your next running route or give you useful advice on how to create the best route for your needs!
          </Text>
          <Text
            style={
              [
                {
                  "fontFamily": "Jakarta-Bold",
                },
                {
                  "fontSize": 18,
                  "marginBottom": 10,
                  "textAlign": "center",
                },
              ]
            }
          >
            How to write to Hadas?
          </Text>
          <Text
            style={
              {
                "fontSize": 14,
                "marginBottom": 10,
                "textAlign": "left",
              }
            }
          >
            You can ask Hadas in natural language, for example:
          </Text>
          <Text
            style={
              {
                "color": "#555",
                "fontSize": 14,
                "fontStyle": "italic",
                "marginBottom": 10,
              }
            }
          >
            - "I would like to run 3km starting from where I am and also end here. Make it easy."
            

            - "I'm not sure how long my running route should be. Any advice?"
            

            - "I haven't run in a while, should I go for a long run or a short one?"
            

          </Text>
          <Text
            style={
              {
                "fontSize": 14,
                "marginBottom": 10,
                "textAlign": "left",
              }
            }
          >
            Hadas will ask you for more details if needed. The more details you provide, the better the recommendations will be!
            

            When the Generate button appears, you can see matching routes! üèÉ‚Äç‚ôÇÔ∏è‚Äç‚û°Ô∏è
          </Text>
          <Text
            style={
              [
                {
                  "fontFamily": "Jakarta-Bold",
                },
                [
                  {
                    "fontSize": 18,
                    "marginBottom": 10,
                    "textAlign": "center",
                  },
                ],
              ]
            }
          >
            Current model: 
            <Text
              style={
                [
                  {
                    "fontFamily": "Jakarta-Bold",
                  },
                  {
                    "color": "#1d4ed8",
                  },
                ]
              }
            >
              AI Model
            </Text>
          </Text>
          <Text
            style={
              [
                {
                  "fontFamily": "Jakarta-Bold",
                },
                {
                  "fontSize": 18,
                  "marginBottom": 10,
                  "textAlign": "center",
                },
              ]
            }
          >
            Current features:
          </Text>
          <View
            style={
              {
                "borderRadius": 8,
                "marginBottom": 1,
                "padding": 10,
                "width": "100%",
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "marginBottom": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Medium",
                    },
                    {
                      "color": "#6b7280",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Running route length
                :
              </Text>
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Bold",
                    },
                    {
                      "color": "#1d4ed8",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                5
              </Text>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "marginBottom": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Medium",
                    },
                    {
                      "color": "#6b7280",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Start location
                :
              </Text>
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Bold",
                    },
                    {
                      "color": "#1d4ed8",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Test Start
              </Text>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "marginBottom": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Medium",
                    },
                    {
                      "color": "#6b7280",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                End location
                :
              </Text>
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Bold",
                    },
                    {
                      "color": "#1d4ed8",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Test End
              </Text>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "marginBottom": 1,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Medium",
                    },
                    {
                      "color": "#6b7280",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Difficulty level
                :
              </Text>
              <Text
                style={
                  [
                    {
                      "fontFamily": "Jakarta-Bold",
                    },
                    {
                      "color": "#1d4ed8",
                    },
                    {
                      "fontSize": 14,
                      "marginBottom": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Medium
              </Text>
            </View>
          </View>
          <button
            onClick={[Function]}
          >
            Reset Chat
          </button>
        </View>
      </RCTScrollView>
      <button
        onClick={[MockFunction]}
      >
        Close
      </button>
    </View>
  </View>
</Modal>
`;
